crates/iceberg-rest-server/src/api/v1/warehouse.rs

```diff
@@ -75,6 +105,113 @@ pub trait WarehouseService<C: Catalog, A: AuthZHandler, S: SecretStore> {
+                let err: IcebergErrorResponse = ErrorModel::builder()
+                    .message("Cannot update 'storage_profile' without new creds.")
+                    .code(StatusCode::CONFLICT.into())
+                    .r#type("CannotUpdateWarehouse")
+                    .build()
+                    .into();
+
+                Err(err)
+            }
+            (None, None) => Ok(()),
+            (Some(new_storage_credential), Some(new_storage_profile)) => {
+                let mut new_storage_profile = new_storage_profile;
+                new_storage_profile
+                    .validate(Some(&new_storage_credential))
+                    .await?;
+
+                let can_be_updated = match (&new_storage_profile, old_storage_profile) {
+                    (StorageProfile::S3(new_profile), StorageProfile::S3(old_profile)) => {
+                        old_profile.can_be_updated_with(new_profile)
+                    }
+                };
+
+                if can_be_updated {
+                    C::update_warehouse_storage_profile(
+                        warehouse_id,
+                        new_storage_profile,
+                        transaction.transaction(),
+                    )
+                    .await?;
+                }
+
+                if let Some(old_cred) = old_storage_credential_id {
+                    S::delete_secret(&old_cred, context.v1_state.secrets.clone()).await?;
+                }
+
+                let new_secret_id =
+                    Some(S::create_secret(new_storage_credential, context.v1_state.secrets).await?);
+
+                Ok(C::update_warehouse_storage_secret_id(
+                    warehouse_id,
+                    new_secret_id,
+                    transaction.transaction(),
+                )
+                .await?)
+            }
+            (Some(new_storage_credential), None) => {
+                if let Some(old_cred) = old_storage_credential_id {
+                    S::delete_secret(&old_cred, context.v1_state.secrets.clone()).await?;
+                }
+
+                let new_secret_id =
+                    Some(S::create_secret(new_storage_credential, context.v1_state.secrets).await?);
+
+                Ok(C::update_warehouse_storage_secret_id(
+                    warehouse_id,
+                    new_secret_id,
+                    transaction.transaction(),
+                )
+                .await?)
+            }
+        }?;
+
+        if let Some(new_status) = request.new_warehouse_status {
+            match new_status {
+                WarehouseStatus::Active => {
+                    C::activate_warehouse(warehouse_id, context.v1_state.catalog)
+                }
+                WarehouseStatus::Inactive => {
+                    C::deactivate_warehouse(warehouse_id, context.v1_state.catalog)
+                }
+            }
+            .await?
+        }
+
+        transaction.commit().await?;
+
+        Ok(UpdateWarehouseResponse {
```

Comment:
We should return unchanged (304) status code if there was no need to change anything.
If credentials where specified, we always need to return 200 even if the old creds are identical to the new creds as we would otherwise leak information about the credentials.